File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    private class Pair{
        String word;
        int dis;

        Pair(String word, int dis){
            this.word = word;
            this.dis = dis;
        }
    }

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> wordSet= new HashSet<>(wordList);

        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(beginWord, 1));
        wordSet.remove(beginWord);

        while(!q.isEmpty()){
            Pair curr = q.poll();

            if(curr.word.equals(endWord)){
                return curr.dis;
            }
            
            for(int i=0; i<curr.word.length(); i++){
                StringBuilder sb = new StringBuilder(curr.word);    
                for(char c='a'; c<='z'; c++){
                    sb.setCharAt(i, c);
                    String check = sb.toString();
                    if(wordSet.contains(check)){
                        q.offer(new Pair(check, curr.dis+1));
                        wordSet.remove(check);
                    }
                }
            }

            
        }

        return 0;
    }
}

File: 0127-word-ladder/0127-word-ladder.py
class Pair:
    def __init__(self, word, dis):
        self.word = word
        self.dis = dis

class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        dictionary = set(wordList)

        q = deque()
        q.append(Pair(beginWord, 1))
        if beginWord in dictionary:
            dictionary.remove(beginWord)

        while q:
            size = len(q)

            for _ in range(size):
                curr = q.popleft()

                if curr.word == endWord:
                    return curr.dis

                for i in range(0, len(curr.word)):
                    for c in string.ascii_lowercase:
                        new_word = curr.word[:i] + c + curr.word[i+1:]
                        if new_word in dictionary:
                            q.append(Pair(new_word, curr.dis + 1))
                            dictionary.remove(new_word)

        return 0





        

        